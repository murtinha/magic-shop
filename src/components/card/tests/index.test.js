import React from 'react';

import { shallow } from 'enzyme';
import Card from '../';

describe('<Card />', () => {
  const onAdd = jest.fn();
  const onRemove = jest.fn();
  const props = {
    card: {
      image: 'path/imagem.png',
      name: 'Black Lotus',
      id: 1,
      manaCost: '0',
      type: 'Artifact',
      rarity: 'Rare',
      setName: 'Magic collection',
      text: 'Card text',
      flavor: 'Card flavor',
      artist: 'murtinha',
    },
    onAdd,
    onRemove,
  };
  const renderShallowWrapper = () => shallow(<Card {...props} />);

  describe('render', () => {
    it('should render properly', () => {
      const wrapper = renderShallowWrapper();
      expect(wrapper).toMatchSnapshot();
    });

    it('should render properly added', () => {
      const wrapper = renderShallowWrapper();
      wrapper.setState({ isAdded: true });
      const div = wrapper.find('.cardFooter');
      expect(div).toMatchSnapshot();
    });

    it('should render properly opened', () => {
      const wrapper = renderShallowWrapper();
      wrapper.setState({ isOpen: true });
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    it('should initialize with correct state', () => {
      const wrapper = renderShallowWrapper();
      expect(wrapper.state()).toEqual({
        isOpen: false,
        isAdded: false,
      });
    });

    it('should set state on image click', () => {
      const wrapper = renderShallowWrapper();
      const img = wrapper.find('.cardBody');
      img.simulate('click');
      expect(wrapper.state().isOpen).toBeTruthy();
    });

    it('should set state on close details', () => {
      const wrapper = renderShallowWrapper();
      wrapper.setState({ isOpen: true });
      expect(wrapper.state().isOpen).toBeTruthy();

      const details = wrapper.find('CardDetails');
      details.simulate('close');
      expect(wrapper.state().isOpen).toBeFalsy();
    });
    
    it('should add card on button click', () => {
      const wrapper = renderShallowWrapper();
      const button = wrapper.find('.btn');
      button.simulate('click');
      expect(wrapper.state().isAdded).toBeTruthy();
      expect(onAdd).toHaveBeenCalledWith({
        name: props.card.name,
        cardId: props.card.id,
      });
    });

    it('should remove card on button click if its added already', () => {
      const wrapper = renderShallowWrapper();
      wrapper.setState({ isAdded: true });
      expect(wrapper.state().isAdded).toBeTruthy();
      const button = wrapper.find('.btn.added');
      button.simulate('click');
      expect(wrapper.state().isAdded).toBeFalsy();
      expect(onRemove).toHaveBeenCalledWith(props.card.id);
    });
  });
}); 
