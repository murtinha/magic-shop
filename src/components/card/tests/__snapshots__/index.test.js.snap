// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> render should render properly 1`] = `
<div
  className="cardWrapper"
>
  <img
    alt="imagem"
    className="cardBody"
    onClick={[Function]}
  />
  <div
    className="cardFooter"
  >
    <button
      className="btn"
      onClick={[Function]}
    >
      Adicionar
    </button>
  </div>
</div>
`;

exports[`<Card /> render should render properly added 1`] = `
<div
  className="cardFooter"
>
  <button
    className="btn added"
    onClick={[Function]}
  >
    Remover
  </button>
</div>
`;

exports[`<Card /> render should render properly opened 1`] = `
<div
  className="cardWrapper"
>
  <img
    alt="imagem"
    className="cardBody"
    onClick={[Function]}
  />
  <CardDetails
    card={
      Object {
        "artist": "murtinha",
        "flavor": "Card flavor",
        "id": 1,
        "image": "path/imagem.png",
        "manaCost": "0",
        "name": "Black Lotus",
        "rarity": "Rare",
        "setName": "Magic collection",
        "text": "Card text",
        "type": "Artifact",
      }
    }
    onClose={[Function]}
  />
  <div
    className="cardFooter"
  >
    <button
      className="btn"
      onClick={[Function]}
    >
      Adicionar
    </button>
  </div>
</div>
`;
